man/arch/craps.6
.\" @(#)craps.6 1.1 92/07/30 SMI; from UCB 4.3 BSD
.TH CRAPS 6 "17 March 1986"
.SH NAME
craps \- the game of craps
.SH SYNOPSIS
.B /usr/games/craps
.SH DESCRIPTION
.I craps
is a form of the game of craps that is played in Las Vegas.
The program simulates the
.IR roller ,
while the user (the
.IR player )
places bets.
The player may choose, at any time, to bet with the roller or with the
.IR House .
A bet of a negative amount is taken as a bet with the House,
any other bet is a bet with the
roller.
.PP
The player starts off with a ``bankroll'' of $2,000.
.PP
The program prompts with:
.IP
bet?
.PP
The bet can be all or part of the player's bankroll.
Any bet over the total bankroll is rejected
and the program prompts with
.B bet?
until a proper bet is made.
.PP
Once the bet is accepted, the
roller throws the dice.
The following rules apply (the player wins or loses depending on
whether the bet is placed with the roller or with the
House; the odds are even).
The
.I first
roll is the roll immediately following a bet:
.IP
1. On the first roll:
.PP
.RS 12
.PD 0
.TP "\w'any\ other\ number\ \ \ 'u"
7\ or\ 11
wins for the roller;
.TP
2,\ 3,\ or\ 12
wins for the House;
.TP
any\ other\ number
is the
.IR point ,
roll again (Rule 2 applies).
.RE
.PD
.IP
2. On subsequent rolls:
.RS 12
.PD 0
.TP "\w'any\ other\ number\ \ \ 'u"
point
roller wins;
.TP
7
House wins;
.TP
any\ other\ number
roll again.
.RE
.PD
.PP
If a player loses the entire bankroll,
the House
will offer to lend the player an additional $2,000.
The program will prompt:
.IP
marker?
.PP
A
.B yes
(or
.BR y )
consummates the loan.
Any other reply terminates the game.
.PP
If a player owes the House money,
the House reminds the player, before a bet is placed,
how many markers are
outstanding.
.PP
If, at any time, the bankroll of a player who has outstanding markers
exceeds $2,000, the House asks:
.IP
Repay marker?
.PP
A reply of
.B yes
(or
.BR y )
indicates the player's willingness to
repay the loan.
If only 1 marker is outstanding, it is immediately repaid.
However, if more than 1 marker are
outstanding, the House asks:
.IP
How many?
.PP
markers the player would like to repay.
If an invalid number is entered
(or just a carriage return),
an appropriate message is printed
and the program will prompt with
.B How\ many?
until a valid number is entered.
.PP
If a player accumulates 10 markers (a total of $20,000 borrowed from the House),
the program informs the player of the situation and exits.
.PP
Should the bankroll of a player who has outstanding markers
exceed $50,000, the
.I total
amount of money borrowed will be
.I automatically
repaid to the House.
.PP
Any player who accumulates $100,000 or more
breaks the bank.
The program then prompts:
.IP
New game?
.PP
to give the House a chance to win back its money.
.PP
Any reply other than
.B yes
is considered to be a
.B no
(except in the case of
.B bet?
or
.BR How\ many? ).
To exit,
send an interrupt (break), \s-1DEL\s+1, or control-\s-1D\s+1.
The program will indicate whether the player won, lost, or broke even.
.SH MISCELLANEOUS
The random number generator for the die numbers uses the seconds from
the time of day.
Depending on system usage, these numbers, at times, may seem strange
but occurrences of this type in a real dice situation are not uncommon.


games/craps.c
#ifndef lint
static	char sccsid[] = "@(#)craps.c 1.1 92/07/30 SMI"; /* from S5R2 1.1 */
#endif

# include	<stdio.h>
# include	<signal.h>

int	lent;
int	minus;
long	h_money;
long	drandno;
char	*get_bet();

static	char	lose[] =	"\nYou lose your bet of $%ld\n";
static	char	win[] =	"\nYou win with %d\n";
static	char	point[] =	"\nYou made your point of %d\n";

main()
{
	register int r;
	register int d;
	register int l;
	int	j;
	long	 money;
	long	patol();
	extern	int finish();
	char	line[BUFSIZ];
	char	*p;
	FILE	*iop;

	iop = stdin;
	h_money = 2000;
	signal(SIGINT,finish);
	time(&drandno);
	printf("Welcome to the Sun floating crap game.\n");
	printf("Commentary? ");
	if ((fgets(line,BUFSIZ,iop) == NULL))
		finish();
	if (line[0] == 'y') {
		printf("\nYou start out with $2,000.  If you  lose  it, the\n");
		printf("House will lend you an additional $2,000. A total\n");
		printf("of $20,000 will be lent.   Should  you  lose  all\n");
		printf("$22,000, you will be escorted from the casino.\n");
		printf("\nYou break the bank with $100,000 or more.\n\n");
	}

	gameloop:
	minus = lent = 0;
	h_money = 2000;
	while (1) {
		l = getdie();
		if (h_money >= 50000 && lent) {
			printf("You now have $%ld\n",h_money);
			printf("You owe the House money, it is ");
			printf("automatically reclaimed\n");
			printf("(The House needs working capital).\n");
			h_money = h_money - (lent * 2000);
			lent = 0;
		}
		if (h_money >= 100000) {
			broke();
			goto gameloop;
		}
		if (h_money <= 0) {
			if (lent == 10) {
				printf("You borrowed $20,000. ");
				printf("Don't you think that's ");
				printf("enough??\nGoodbye\n");
				exit (1);
			}
			printf("marker? ");
			fgets(line,BUFSIZ,iop);
			if (line[0] == 'y') {
				h_money = 2000;
				lent++;
			}
			else finish();
		}
		if (lent) {
			printf("You have %d marker(s) outstanding\n",lent);
			if (h_money > 2000) {
				printf("You now have $%ld\n",h_money);
				printf("Repay marker? ");
				if ((fgets(line,BUFSIZ,iop) == NULL))
					finish();
				if (line[0] == '\n' || line[0] == 'n')
					;
				else if (line[0] == 'y') {
					if (lent == 1) {
						h_money = h_money - 2000;
						lent--;
					}
					else how_many();
				}
			}
		}
		printf("Your bankroll is $%ld\n",h_money);
		if (p = get_bet())
			money = patol(p);
		else money = 2000000;
		while (money > h_money) {
			if (minus)
				--minus;
			printf("Illegal bet\n");
			if (p = get_bet())
				money = patol(p);
			else money = 2000000;
		}
		d = getdie();
		printf ("%d\t%d\n",l,d);
		r = d + l;
		if (r == 7 || r == 11) {
			if (minus) {
				printf(lose,money);
				h_money = h_money - money;
				--minus;
			}
			else {
				printf(win,r);
				h_money = h_money + money;
			}
		}
		else if (r == 2 || r == 3 || r == 12) {
			if (minus) {
				printf(win,r);
				h_money = h_money + money;
				--minus;
			}
			else {
				printf(lose,money);
				h_money = h_money - money;
			}
		}
		else {
			printf("The point is %d\n",r);
			j = 0;
			while (j != r) {
				l = getdie();
				d = getdie();
				printf ("\t%d\t%d\n",l,d);
				j = d + l;
				if (j == 7) {
					if (minus) {
						printf(win,j);
						h_money = h_money + money;
						--minus;
					}
					else {
						printf(lose,money);
						h_money = h_money - money;
					}
					break;
				}
			}
			if (j == r) {
				if (minus) {
					printf(lose,money);
					h_money = h_money - money;
					--minus;
				}
				else {
					printf(point,j);
					h_money = h_money + money;
				}
			}
		}
	}
}


getdie()
{
	register	int retval;

	do {
		drandno *= 2345699l;
		drandno += 12345701l;
	} while (!(retval = drandno % 7));
	if (retval < 0)
		retval += 7;
	return(retval);
}


finish()
{
	long	 money;

	money = ((lent * 2000) + 2000);
	h_money = h_money - money;
	if (h_money > 0)
		printf("\nYou win $%ld\n",h_money);
	else if (h_money == 0)
		printf("\nYou break even\n");
	else {
		h_money = (-h_money);
		printf("\nYou lose $%ld\n",h_money);
	}
	exit (1);
}


broke()
{
	char	line[BUFSIZ];
	FILE	*iop;

	iop = stdin;
	printf("\nYour luck is pretty good.  You broke the bank\n");
	printf("New game? ");
	fgets(line,BUFSIZ,iop);
	if (line[0] == 'y')
		return;
	exit (1);
}


a_number(l)
char	*l;
{
	register int j;

	for (j = 0; l[j] != '\0'; j++)
		if ('0' <= l[j] && l[j] <= '9')
			continue;
		else return(0);
	if (l[0] == '\0')
		return(0);
	return(1);
}


how_many()
{
	FILE *iop;
	register int repay;
	char	line[BUFSIZ];

	iop = stdin;
	repay = 0;
	while (repay <= lent) {
		printf("How many? ");
		if ((fgets(line,BUFSIZ,iop) == NULL)) {
			printf("\nYou refused to pay\n");
			finish();
		}
		line[strlen(line) - 1] = '\0';
		if (a_number(line)) {
			repay = atoi(line);
			if (repay > lent) {
				printf("You don't have that many markers\n");
				repay = 0;
				continue;
			}
			if (repay * 2000 > h_money) {
				printf("You don't have enough money\n");
				repay = 0;
				continue;
			}
			if (repay * 2000 == h_money) {
				printf("That wouldn't leave you much!!\n");
				repay = 0;
				continue;
			}
			h_money = h_money - (repay * 2000);
			lent = lent - repay;
			printf("You repayed %d marker(s)\n",repay);
			break;
		}
		else {
			printf("Illegal entry\n");
			repay = 0;
		}
	}
	return;
}


char	*get_bet()
{
	static char line[BUFSIZ];
	register char *p;
	FILE *iop;

	iop = stdin;
	printf("bet? ");
	if ((fgets(line,BUFSIZ,iop)) == NULL)
		finish();
	if (line[0] != '\n') {
		line[strlen(line) - 1] = '\0';
		p = line;
		if (line[0] == '-') {
			p++;
			minus++;
		}
		if (!a_number(p))
			return(0);
		else return(p);
	}
	else return(0);
}


long patol(s)
register char *s;
{
	long i;

	i = 0;
	while (*s >= '0' && *s <= '9')
		i = 10*i + *s++ - '0';

	if (*s)
		return(-1);
	return(i);
}
